src/qrcode/QrCode.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ *  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "QrCode.h"
+#include "QrCode_p.h"
+
+#include <QBrush>
+#include <QByteArray>
+#include <QImage>
+#include <QPainter>
+#include <QPen>
+#include <QString>
+#include <QVariant>
+
+QrCodePrivate::QrCodePrivate()
+    : m_qrcode(nullptr)
+{
+}
+
+QrCodePrivate::~QrCodePrivate()
+{
+    if (m_qrcode) {
+        QRcode_free(m_qrcode);
+    }
+}
+
+QrCode::QrCode()
+    : d_ptr(new QrCodePrivate())
+{
+}
+
+QrCode::QrCode(const QString& data, const Version version, const ErrorCorrectionLevel ecl, const bool caseSensitive)
+    : d_ptr(new QrCodePrivate())
+{
+    init(data, version, ecl, caseSensitive);
+}
+
+QrCode::QrCode(const QByteArray& data, const Version version, const ErrorCorrectionLevel ecl)
+    : d_ptr(new QrCodePrivate())
+{
+    init(data, version, ecl);
+}
+
+QrCode::~QrCode() = default;
+
+void QrCode::init(const QString& data, const Version version, const ErrorCorrectionLevel ecl, bool caseSensitive)
+{
+    if (data.isEmpty()) {
+        return;
+    }
+
+    d_ptr->m_qrcode = QRcode_encodeString(data.toLocal8Bit().data(),
+                                          static_cast<int>(version),
+                                          static_cast<QRecLevel>(ecl),
+                                          QR_MODE_8,
+                                          caseSensitive ? 1 : 0);
+}
+
+void QrCode::init(const QByteArray& data, const Version version, const ErrorCorrectionLevel ecl)
+{
+    if (data.isEmpty()) {
+        return;
+    }
+
+    d_ptr->m_qrcode = QRcode_encodeData(data.size(),
+                                        reinterpret_cast<const unsigned char*>(data.data()),
+                                        static_cast<int>(version),
+                                        static_cast<QRecLevel>(ecl));
+}
+
+QVariant QrCode::toQImage(const int size, const int margin) const
```

Comment:
I've been thinking about this. Qt has a `QSvgGenerator` class that works with `QPainter`. Could you use it instead of `QImage` and make the output to be of some *physical* size taking the screen DPI into account instead of hardcoding the image to 200x200 pixels. The actual vector image can have 1x1 rectangles for each dot but the scaling should be external.

A bonus feature would be that the vector image would scale with the window so you can adjust it.

Reply:
I don't know if using an SVG would be better. Can you review the changes I've made to address this observation?

Reply:
The change to pre-calculate the size before is a definite improvement!

If you're going to keep it as a bitmap, it would make sense to prevent resizing the window as it doesn't serve a useful function.

The change to SVG/vector would hopefully allow the image to automatically resize with the window if properly implemented and you would never need to repaint it manually. That was the main point. Setting the physical size of the bitmap based on DPI is more important than that, though.

Reply:
> If you're going to keep it as a bitmap, it would make sense to prevent resizing the window as it doesn't serve a useful function.

Agreed.

Reply:
Although, since QtSVG is already a dependency of KeePassXC, I guess we could try the approach you're proposing and render the SVG using `QGraphicsSvgItem`.

Reply:
@hifi I will explore the SVG stuff in a next PR if you're OK with it. I don't have more time left in the next couple of days, and they're feature-freezing 2.3.0 soon.